File: rxbinding-appcompat-v7/src/main/java/com/jakewharton/rxbinding2/support/v7/widget/SearchViewQueryTextChangesObservable.java
package com.jakewharton.rxbinding2.support.v7.widget;

import android.support.v7.widget.SearchView;
import com.jakewharton.rxbinding2.InitialValueObservable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class SearchViewQueryTextChangesObservable extends InitialValueObservable<CharSequence> {
  private final SearchView view;

  SearchViewQueryTextChangesObservable(SearchView view) {
    this.view = view;
  }

  @Override protected void subscribeListener(Observer<? super CharSequence> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.setOnQueryTextListener(listener);
  }

  @Override protected CharSequence getInitialValue() {
    return view.getQuery();
  }

  final class Listener extends MainThreadDisposable implements SearchView.OnQueryTextListener {
    private final SearchView searchView;
    private final Observer<? super CharSequence> observer;

    Listener(SearchView searchView, Observer<? super CharSequence> observer) {
      this.searchView = searchView;
      this.observer = observer;
    }

    @Override public boolean onQueryTextChange(String s) {
      if (!isDisposed()) {
        observer.onNext(s);
        return true;
      }
      return false;
    }

    @Override public boolean onQueryTextSubmit(String query) {
      return false;
    }

    @Override protected void onDispose() {
      searchView.setOnQueryTextListener(null);
    }
  }
}

File: rxbinding-appcompat/src/main/java/com/jakewharton/rxbinding3/appcompat/SearchViewQueryTextChangesObservable.kt
@file:JvmName("RxSearchView")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.appcompat

import androidx.annotation.CheckResult
import androidx.appcompat.widget.SearchView
import com.jakewharton.rxbinding3.InitialValueObservable
import com.jakewharton.rxbinding3.internal.checkMainThread
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

/**
 * Create an observable of character sequences for query text changes on `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Note:* A value will be emitted immediately on subscribe.
 */
@CheckResult
fun SearchView.queryTextChanges(): InitialValueObservable<CharSequence> {
  return SearchViewQueryTextChangesObservable(this)
}

private class SearchViewQueryTextChangesObservable(
  private val view: SearchView
) : InitialValueObservable<CharSequence>() {

  override fun subscribeListener(observer: Observer<in CharSequence>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.setOnQueryTextListener(listener)
  }

  override val initialValue get() = view.query

  private class Listener(
    private val searchView: SearchView,
    private val observer: Observer<in CharSequence>
  ) : MainThreadDisposable(), SearchView.OnQueryTextListener {

    override fun onQueryTextChange(s: String): Boolean {
      if (!isDisposed) {
        observer.onNext(s)
        return true
      }
      return false
    }

    override fun onQueryTextSubmit(query: String): Boolean {
      return false
    }

    override fun onDispose() {
      searchView.setOnQueryTextListener(null)
    }
  }
}
